experimental/zerotrie/src/byte_phf/mod.rs

```diff
@@ -0,0 +1,362 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+#[cfg(feature = "alloc")]
+mod builder;
+#[cfg(feature = "alloc")]
+mod cached_owned;
+
+#[cfg(feature = "alloc")]
+pub use builder::find;
+#[cfg(feature = "alloc")]
+pub use cached_owned::PerfectByteHashMapCacheOwned;
+
+const P_FAST_MAX: u8 = 11;
+const Q_FAST_MAX: u8 = 95;
+
+#[cfg(feature = "alloc")] // used in the builder code
+const P_REAL_MAX: u8 = 15;
+#[cfg(feature = "alloc")] // used in the builder code
+const Q_REAL_MAX: u8 = 127;
+
+/// Like slice::split_at but returns an Option instead of panicking
+#[inline]
+fn debug_split_at(slice: &[u8], mid: usize) -> Option<(&[u8], &[u8])> {
+    if mid > slice.len() {
+        debug_assert!(false, "debug_split_at: index expected to be in range");
+        None
+    } else {
+        // Note: We're trusting the compiler to inline this and remove the assertion
+        // hiding on the top of slice::split_at: `assert(mid <= self.len())`
+        Some(slice.split_at(mid))
+    }
+}
+
+#[inline]
+fn debug_get(slice: &[u8], index: usize) -> Option<u8> {
+    match slice.get(index) {
+        Some(x) => Some(*x),
+        None => {
+            debug_assert!(false, "debug_get: index expected to be in range");
+            None
+        }
+    }
+}
+
+/// Invariant: n > 0
+#[inline]
+pub fn f1(byte: u8, p: u8, n: usize) -> usize {
```

Comment:
issue: docs on all these functions please

Reply:
and inside them

Reply:
Done

Reply:
Done